lib/iris/_data_manager.py

```diff
@@ -185,7 +199,15 @@ def _assert_axioms(self):
                 'real data and realised {!r}.')
         assert state, emsg.format(self._realised_dtype)
 
-    def _deepcopy(self, memo, data=None, realised_dtype='none'):
+        state = not (self.has_lazy_data() and
+                     self._lazy_array.dtype.kind != 'f' and
+                     self._realised_dtype is not None)
```

Comment:
I agree is pretty hard to read, I have that problem also, and I wrote it ...

The intent is that it is an invalid state to have "lazy data that isn't float but there is a realised dtype".

The logic does indeed capture that ... and we need to negate the logic as `assert` only fires when it's `False` ... so I believe this is right!